src/Dapr.Messaging/PublishSubscribe/ConnectionManager.cs

```diff
@@ -0,0 +1,148 @@
+﻿// ------------------------------------------------------------------------
+// Copyright 2024 The Dapr Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//     http://www.apache.org/licenses/LICENSE-2.0
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// ------------------------------------------------------------------------
+
+using System.Threading.Channels;
+using Grpc.Core;
+using C = Dapr.AppCallback.Autogen.Grpc.v1;
+using P = Dapr.Client.Autogen.Grpc.v1;
+
+namespace Dapr.Messaging.PublishSubscribe;
+
+/// <summary>
+/// Maintains the streaming connection to the Dapr sidecar so it can be repurposed without
+/// multiple callers opening separate connections.
+/// </summary>
+internal sealed class ConnectionManager : IAsyncDisposable
```

Comment:
This seems to get reused for each `PublishSubscribeReceiver`, right?  I don't think that should be so, if I understand the semantics of the `SubscribeTopicEventsAlpha1()` API. When called, the method returns a bi-directional stream.  The Dapr runtime expects that the very first "request" sent from the caller to be the topic to subscribe to, and then the remainder of the "requests" will be message acknowledgements (until either side closes the stream).

To subscribe to multiple topics in parallel, or to subscribe after a previous subscription has completed, the client must create a new stream (i.e. another call to `SubscribeTopicEventsAlpha1()`) for each subscription, right?

Reply:
In light of @JoshVanL 's confirmation of @philliphoff 's read of the spec, I've gone ahead and made an update to remove the `ConnectionManager` altogether. Instead, the following now happens:

Each call to `DaprPublishSubscribeClient.Register` (via `DaprPublishSubscribeGrpcClient`) creates a new `PublishSubscribeReceiver` for the indicated pubsub component name, topic name, options, and handling delegate. Following a call to `SubscribeAsync` with a cancellation token, This validates that the subscription hasn't already been initialized (to prevent the initial request from being submitted twice), then does the following:
1) Sends the initial request via `SubscribeTopicEventsRequestInitialAlpha1` and sets up an asynchronous reader of inbound messages to pass into the topic message `Channel` writer.
2) Sets up an asynchronous reader of the acknowledgements `Channel` reader holding each of the topic acknowledgements (each containing a message ID and `TopicEventResponseStatus` reflecting the action to take on each message).
3) Sets up an asynchronous reader of the topic message `Channel` reader holding each of the messages to pass each into the delegate `TopicMessageHandler` as provided to be evaluated.

As each message is evaluated, either an action is returned or it times out and a default configured action is paired with the message ID and persisted back to the acknowledgements `Channel`. As these "acknowledgements" are received, they're sent back over the persisted connection to the sidecar.

I refactored this a little bit so that if the operation is cancelled or disposed, it attempts first to complete and drain the inbound message `Channel` then complete and drain the acknowledgement `Channel`.

No changes needed to the example project already written.

Reply:
> How should the underlying connection be shared if it cannot receive more than one initial request?

I assume @JoshVanL is referring to the underlying gRPC connection (i.e. in .NET, `GrpcChannel`)?